% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rnaseq_workflow_qc.R
\name{load_paired_end_qc_data}
\alias{load_paired_end_qc_data}
\title{Load QC metrics data from paired-end RNA-seq processing}
\usage{
load_paired_end_qc_data(data_dir)
}
\arguments{
\item{data_dir}{A string denoting the local path where all the data files are stored.}
}
\value{
A list with several elements:
\itemize{
\item \code{fastqc}: Data frame imported from \code{multiqc_fastqc.txt}.
\item \code{hisat2}: Data frame imported from \code{multiqc_hisat2.txt}.
\item \code{rseqc_bam}: Data frame imported from \code{multiqc_rseqc_bam_stat.txt}.
\item \code{rseqc_alignment_category}: Data frame imported from \code{multiqc_rseqc_read_distribution.txt}.
\item \code{salmon}: Data frame imported from \code{multiqc_salmon.txt}.
\item \code{trimmomatic}: Data frame imported from \code{multiqc_trimmomatic.txt}.
\item \code{inner_dist}: Data frame imported from \code{rseqc_inner_distance_plot.tsv}.
\item \code{frag_length}: Data frame imported from \code{salmon_plot.tsv}.
\item \code{phred_bp}: Data frame imported from \code{fastqc_per_base_sequence_quality_plot.tsv}.
\item \code{phred_seq}: Data frame imported from \code{fastqc_per_sequence_quality_scores_plot.tsv}.
\item \code{gc_content}: Data frame imported from \code{fastqc_per_sequence_gc_content_plot.tsv}.
\item \code{seq_duplication}: Data frame imported from \code{fastqc_sequence_duplication_levels_plot.tsv}.
\item \code{adapter_content}: Data frame imported from \code{fastqc_adapter_content_plot.tsv}.
\item \code{known_junction}: Data frame imported from \code{rseqc_known_junction_saturation_plot.tsv}.
\item \code{novel_junction}: Data frame imported from \code{rseqc_novel_junction_saturation_plot.tsv}.
}
}
\description{
Load QC data files generated by MultiQC for various stages of paired-end RNA-seq data
processing using the automated RNA-seq data processing WDL workflow.
}
\details{
The automated RNA-seq data processing workflow produces several tables with various metrics.
Additionally, several tables can be exported from the MultiQC HTML report. This function will
import all of these data into a list for further processing in R. The assumed data files are
\itemize{
\item \code{multiqc_fastqc.txt}: An auto-generated workflow file with general FASTQC statistics.
\item \code{multiqc_hisat2.txt}: An auto-generated workflow file with HISAT2 alignment statistics.
\item \code{multiqc_rseqc_bam_stat.txt}: An auto-generated workflow file with alignment statistics
generated by RSeQC.
\item \code{multiqc_rseqc_read_distribution.txt}: An auto-generated workflow file with alignment
categorizations to different types of genomic regions (introns, exons, integenic, etc.).
\item \code{multiqc_salmon.txt}: An auto-generated workflow file with mapping statistics from Salmon.
\item \code{multiqc_trimmomatic.txt}: An auto-generated workflow file with trimming and read quality
filtering statistics using Trimmomatic.
\item \code{rseqc_inner_distance_plot.tsv}: Table exported from the MultiQC HTML report with inner
distance statistics between pairs of reads.
\item \code{salmon_plot.tsv}: Table exported from the MultiQC HTML report with RNA fragment length
distribution statistics as estimated by Salmon.
\item \code{fastqc_per_base_sequence_quality_plot.tsv}: Table exported from the MultiQC HTML report with
per base PHRED scores statistics generated by FASTQC.
\item \code{fastqc_per_sequence_quality_scores_plot.tsv}: Table exported from the MultiQC HTML report
with per sequence PHRED score statistics generated by FASTQC.
\item \code{fastqc_per_sequence_gc_content_plot.tsv}: Table exported from the MultiQC HTML report with
per sequence GC content statistics generated by FASTQC.
\item \code{fastqc_sequence_duplication_levels_plot.tsv}: Table exported from the MultiQC HTML report
with sequence duplication statistics generated by FASTQC.
\item \code{fastqc_adapter_content_plot.tsv}: Table exported from the MultiQC HTML report with adapter
content statistics generated by FASTQC.
\item \code{rseqc_known_junction_saturation_plot.tsv}: Table exported from the MultiQC HTML report with
junction saturation statistics for known splice junctions.
\item \code{rseqc_novel_junction_saturation_plot.tsv}: Table exported from the MultiQC HTML report with
junction saturation statistics for novel splice junctions.
}
}
