% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/rnaseq_workflow_qc.R
\name{plot_unique_read_pct}
\alias{plot_unique_read_pct}
\title{Unique read percentage per sample scatterplot}
\usage{
plot_unique_read_pct(
  data,
  group = NULL,
  point_size = 4,
  fill = NULL,
  alpha = 0.75,
  sort = T,
  ids = NULL,
  invert = F,
  return_data = F
)
}
\arguments{
\item{data}{A data frame with sequencing depth data in the same format as \code{fastqc} returned
by \verb{\link{load_paired_end_qc_data}}.}

\item{group}{A single column data frame denoting categorical variable values to use to
partition the data points into separate distributions/boxplots. Row names should match rownames
in \code{data}.}

\item{point_size}{A numeric. The size for data points.}

\item{fill}{A string. The fill color for data points.}

\item{alpha}{A numeric. The alpha level for data points.}

\item{sort}{A logical. Should the x-axis be ordered by sequencing depth?}

\item{ids}{A vector of rownames for subsetting \code{data} for plotting.}

\item{invert}{A logical. Should \code{ids} be used for excluding rows from plotting instead?}

\item{return_data}{A logical. Should plot data be returned instead of a ggplot object?}
}
\value{
A ggplot object.

A ggplot object unless \code{return_data} is \code{TRUE}, then a data frame with the
unique read percentage and grouping variable.
}
\description{
Create a unique read percentage by sample scatterplot.
}
\seealso{
\verb{\link{load_paired_end_qc_data}}
}
